fusequery/query/src/optimizers/optimizer_projection_push_down.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use anyhow::Result;
+use common_arrow::arrow::error::Result as ArrowResult;
+use common_datavalues::DataField;
+use common_datavalues::DataSchema;
+use common_datavalues::DataSchemaRef;
+use common_planners::AggregatorFinalPlan;
+use common_planners::AggregatorPartialPlan;
+use common_planners::ExpressionPlan;
+use common_planners::FilterPlan;
+use common_planners::PlanNode;
+use common_planners::ProjectionPlan;
+use common_planners::ReadDataSourcePlan;
+use common_planners::SortPlan;
+
+use crate::optimizers::IOptimizer;
+use crate::optimizers::OptimizerCommon;
+use crate::sessions::FuseQueryContextRef;
+
+pub struct ProjectionPushDownOptimizer {}
+
+impl ProjectionPushDownOptimizer {
+    pub fn create(_ctx: FuseQueryContextRef) -> Self {
+        ProjectionPushDownOptimizer {}
+    }
+}
+
+/// Recursively walk an expression tree, collecting the unique set of column names
+/// referenced in the expression
+fn expr_to_column_names(expr: &ExpressionPlan, accum: &mut HashSet<String>) -> Result<()> {
+    let expressions = OptimizerCommon::expression_plan_children(expr)?;
+
+    let _expressions = expressions
+        .iter()
+        .map(|e| expr_to_column_names(e, accum))
+        .collect::<Result<Vec<_>>>()?;
+
+    if let ExpressionPlan::Column(name) = expr {
+        accum.insert(name.clone());
+    }
+    Ok(())
+}
+
+/// Recursively walk a list of expression trees, collecting the unique set of column
+/// names referenced in the expression
+fn exprvec_to_column_names(expr: &[ExpressionPlan], accum: &mut HashSet<String>) -> Result<()> {
+    for e in expr {
+        expr_to_column_names(e, accum)?;
+    }
+    Ok(())
+}
+
+fn expr_to_name(e: &ExpressionPlan) -> Result<String> {
+    match e {
+        ExpressionPlan::Column(name) => Ok(name.clone()),
+        _ => Err(anyhow::anyhow!("Ignore ExpressionPlan that is not Column."))
+    }
+}
+
+fn get_projected_schema(
```

Comment:
Not sure these private function may be used by other optimizer rule in future?
1) If it is, it would be better move to optimizer_common.rs
2) if not, it's ok here

Reply:
Sure. We may leave it here for the moment and keep an eye on them.